### PR TITLE
Build filter executor from arrow ScanOptions' filter expression

### PR DESCRIPTION
**Pull request by eddyxu**

### CODE REVIEW

File: cpp/src/lance/io/filter_test.cc [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+                                           ::arrow::field("label", ::arrow::utf8())}));
+
+TEST_CASE("Test without filter") {
+  auto empty_filter =
+      lance::io::Filter::Make(kSchema, ::arrow::compute::literal(true)).ValueOrDie();
+  CHECK(empty_filter == nullptr);
+}
+
+TEST_CASE("Test with one condition") {
+  auto filter = lance::io::Filter::Make(kSchema, equal(field_ref("value"), literal("32")));
```
Comment (changhiskhan) [inline cpp/src/lance/io/filter_test.cc:45]: Ok this just checks the filter construction and doesn't actually apply the filter right

File: cpp/src/lance/io/filter.h [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+#include <tuple>
+
+#include "lance/format/schema.h"
+
+namespace lance::io {
+
+class FileReader;
+
+/// Filter.
+class Filter {
```
Comment (changhiskhan) [inline cpp/src/lance/io/filter.h:30]: Just curious how similar this is to Parquet? Do you have a clearer idea on where to draw the boundary between what's in the format vs compute layer at this point?

File: cpp/src/lance/io/filter.cc [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    /// All scalar?
+    return nullptr;
+  }
+  auto field_refs = ::arrow::compute::FieldsInExpression(filter);
+  std::vector<std::string> columns;
+  for (auto& ref : field_refs) {
+    columns.emplace_back(std::string(*ref.name()));
+  }
+  ARROW_ASSIGN_OR_RAISE(auto filter_schema, schema.Project(columns));
+  return std::unique_ptr<Filter>(new Filter(filter_schema, filter));
```
Comment (changhiskhan) [inline cpp/src/lance/io/filter.cc:40]: I forgot, how does this get wrapped into Result?

**Review by changhiskhan — APPROVED** [review]

File: cpp/src/lance/io/filter_test.cc [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+#include "lance/arrow/stl.h"
+#include "lance/arrow/type.h"
+#include "lance/format/schema.h"
+
+using ::arrow::compute::equal;
+using ::arrow::compute::field_ref;
+using ::arrow::compute::literal;
+using ::arrow::compute::or_;
+
+const auto kSchema =
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply cpp/src/lance/io/filter_test.cc:33]: `k` is the constant prefix, per [Google C++ Style](https://google.github.io/styleguide/cppguide.html#Constant_Names) which is followed by arrow as well.

File: cpp/src/lance/io/filter.h (hunk 2 above)
Comment (eddyxu) [reply cpp/src/lance/io/filter.h:30]: The current plan is that, format / storage api covers single data source project + filter (including index based filter as well).  So the pylance API and pyarrow API can pass enough filters to accelerate I/Os.
Do not plan to cover things like computation, join, catelog and etc.

This is different to parquet. For parquet, there is no row-level filtering (i.e., here we are using the mask created from the filter to fetch other columns). In parquet, all the projected column/row groups are read into memory before applying filters.

File: cpp/src/lance/io/filter.cc (hunk 3 above)
Comment (eddyxu) [reply cpp/src/lance/io/filter.cc:40]: so `Result<T>` has two constructors, one from `Status`, one from `T`. 
Here T = `std::unique_ptr<Filter>`, so it `std::move`-ed the filter into `Result`.

File: cpp/src/lance/io/filter_test.cc (hunk 1 above)
Comment (eddyxu) [reply cpp/src/lance/io/filter_test.cc:45]: yes, this is just creating filters

**Pull request merged**